[SPARK-46216][CORE] Improve `FileSystemPersistenceEngine` to support compressions

dongjoon-hyun opened the pull request:
### What changes were proposed in this pull request?

This PR aims to improve `FileSystemPersistenceEngine` to support compressions via a new configuration, `spark.deploy.recoveryCompressionCodec`.

### Why are the changes needed?

To allow the users to choose a proper compression codec for their workloads. For `JavaSerializer` case, `LZ4` compression is **2x** faster than the baseline (no compression).
```
OpenJDK 64-Bit Server VM 17.0.9+9-LTS on Linux 5.15.0-1051-azure
AMD EPYC 7763 64-Core Processor
2000 Workers:                                             Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
----------------------------------------------------------------------------------------------------------------------------------------
ZooKeeperPersistenceEngine with JavaSerializer                     2276           2360         115          0.0     1137909.6       1.0X
ZooKeeperPersistenceEngine with KryoSerializer                     1883           1906          34          0.0      941364.2       1.2X
FileSystemPersistenceEngine with JavaSerializer                     431            436           7          0.0      215436.9       5.3X
FileSystemPersistenceEngine with JavaSerializer (lz4)               209            216           9          0.0      104404.1      10.9X
FileSystemPersistenceEngine with JavaSerializer (lzf)               199            202           2          0.0       99489.5      11.4X
FileSystemPersistenceEngine with JavaSerializer (snappy)            192            199           9          0.0       95872.9      11.9X
FileSystemPersistenceEngine with JavaSerializer (zstd)              258            264           6          0.0      129249.4       8.8X
FileSystemPersistenceEngine with KryoSerializer                     139            151          13          0.0       69374.5      16.4X
FileSystemPersistenceEngine with KryoSerializer (lz4)               159            165           8          0.0       79588.9      14.3X
FileSystemPersistenceEngine with KryoSerializer (lzf)               180            195          18          0.0       89844.0      12.7X
FileSystemPersistenceEngine with KryoSerializer (snappy)            164            183          18          0.0       82016.0      13.9X
FileSystemPersistenceEngine with KryoSerializer (zstd)              206            218          11          0.0      102838.9      11.1X
BlackHolePersistenceEngine                                            0              0           0         35.1          28.5   39908.5X
```

### Does this PR introduce _any_ user-facing change?

No, this is a new feature.

### How was this patch tested?

Pass the CIs with the newly added test case.

### Was this patch authored or co-authored using generative AI tooling?

No.

dongjoon-hyun committed:
[SPARK-46216][CORE] Improve `FileSystemPersistenceEngine` to support compressions

dongjoon-hyun committed:
Add result

HyukjinKwon commented on core/src/main/scala/org/apache/spark/internal/config/Deploy.scala:
```diff
@@ -39,6 +41,14 @@ private[spark] object Deploy {
     .createWithDefault(RecoverySerializer.JAVA.toString)
 
+  val RECOVERY_COMPRESSION_CODEC = ConfigBuilder("spark.deploy.recoveryCompressionCodec")
+    .doc("A compression codec for persistence engines. lz4 (default), lzf, snappy, and zstd. " +
```
How the end users specify no compression codec? Seems like we would need sth like `none` and/or `uncompressed`.

HyukjinKwon reviewed (approved):
Seems fine to me

dongjoon-hyun committed:
Address comments

dongjoon-hyun commented:
Thank you so much. I addressed your comment via converting `spark.deploy.recoveryCompressionCodec` as `optionalString`. And, the default is no compression.

dongjoon-hyun commented:
Merged to master for Apache Spark 4.

sadikovi commented:
@dongjoon-hyun Could you check if `SPARK-46216: FileSystemPersistenceEngine with compression` passes on Java 8? I have encountered an error when running it on Java 8.

```
*** 1 TEST FAILED ***
PersistenceEngineSuite:

- SPARK-46216: FileSystemPersistenceEngine with compression *** FAILED *** (660 milliseconds)
  java.io.IOException: StreamClosed
  at com.github.luben.zstd.ZstdOutputStreamNoFinalizer.flush(ZstdOutputStreamNoFinalizer.java:377)
  at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:141)
  at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
  at org.apache.spark.deploy.master.FileSystemPersistenceEngine.$anonfun$serializeIntoFile$3(FileSystemPersistenceEngine.scala:76)
  at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally(SparkErrorUtils.scala:73)
  at org.apache.spark.util.SparkErrorUtils.tryWithSafeFinally$(SparkErrorUtils.scala:61)
  at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:103)
  at org.apache.spark.deploy.master.FileSystemPersistenceEngine.serializeIntoFile(FileSystemPersistenceEngine.scala:72)
  at org.apache.spark.deploy.master.FileSystemPersistenceEngine.persist(FileSystemPersistenceEngine.scala:47)
  at org.apache.spark.deploy.master.PersistenceEngineSuite.testPersistenceEngine(PersistenceEngineSuite.scala:103)
``` 

Seems Zstd already closes the stream so calling fileOut.close() causes an exception to be thrown.

dongjoon-hyun commented:
Hi, @sadikovi . 

If you are in the middle of backporting, please follow the merge order.

I believe you need this first.
- #44113

dongjoon-hyun commented:
According to your error message, I guess this one is what you need, @sadikovi .
- https://github.com/apache/spark/pull/44113/files#r1412659395

sadikovi commented:
Thanks. I actually have that change as I am checking out master (I confirmed the code is there). Let me double check again on Ubuntu.
